src/transforms/bin.js

```diff
@@ -305,38 +341,66 @@ function isInterval(t) {
   return t ? typeof t.range === "function" : false;
 }
 
-function binset(bin) {
-  return [bin, new Set(bin)];
-}
-
-function bincumset([bin], j, bins) {
-  return [
-    bin,
-    {
-      get size() {
-        for (let k = 0; k <= j; ++k) {
-          if (bins[k][1].size) {
-            return 1; // a non-empty value
+function Bin(EX, EY) {
+  return EX && EY
+    ? function* (I) {
+        const X = EX.bin(I); // first bin on x
+        for (const [ix, [x1, x2]] of EX.entries()) {
+          const Y = EY.bin(X[ix]); // then bin on y
+          for (const [iy, [y1, y2]] of EY.entries()) {
+            yield [Y[iy], {x1, y1, x2, y2}];
           }
         }
-        return 0;
-      },
-      has(i) {
-        for (let k = 0; k <= j; ++k) {
-          if (bins[k][1].has(i)) {
-            return true;
-          }
+      }
+    : EX
+    ? function* (I) {
+        const X = EX.bin(I);
+        for (const [i, [x1, x2]] of EX.entries()) {
+          yield [X[i], {x1, x2}];
         }
-        return false;
       }
-    }
-  ];
+    : function* (I) {
+        const Y = EY.bin(I);
+        for (const [i, [y1, y2]] of EY.entries()) {
+          yield [Y[i], {y1, y2}];
+        }
+      };
 }
 
-function binfilter([{x0, x1}, set]) {
-  return [x0, x1, set.size ? (I) => I.filter(set.has, set) : binempty];
+// non-cumulative distribution
+function bin1(E, T, V) {
```

Comment:
Here's a bin quantizer that can be used in lieu of bin1 if we use the "number" ticks. However, due to floating-point rounding, we need to undershoot then correct course… this is really not your most beautiful piece of code—and I fear it might create more problems. It also doesn't handle the degenerate case

```
function binq(E, T, V) {
  if (T.length < 2) return bin1(E, T, V); // degenerate case
  const a = T[0];
  const b = (1 + 1e-12) / (T[1] - T[0]);
  return (I) => {
    const B = E.map(() => []);
    for (const i of I) {
      let j = Math.floor(b * (V[i] - a));
      if (T[j] > V[i]) j++;
      B[j]?.push(i);
    }
    return B;
  };
}
```

In my tests it's about 30% faster, so maybe worth a shot (later).

Reply:
I think we’ll want to use the value returned by tickIncrement directly here (like we do in d3.bin) rather than “rediscovering” it as T[1] - T[0]. But yes I suggest we defer this optimization to later.